internal/dag/conditions.go

```diff
@@ -22,37 +22,66 @@ import (
 	contour_api_v1 "github.com/projectcontour/contour/apis/projectcontour/v1"
 )
 
-// mergePathMatchConditions merges the given slice of prefix MatchConditions into a single
-// prefix Condition.
+// mergePathMatchConditions merges the given slice of prefix or exact MatchConditions into a single
+// prefix/exact Condition.
+// If the root condition is prefix, and included condition is prefix then after merging we get one prefix condition.
+// If the root condition is prefix, and included condition is exact then after merging we get one exact condition.
+// If the root condition is exact, and included condition is exact then after merging we get one exact condition.
+// If the root condition is exact, and included condition is prefix then after merging we get one prefix condition.
+// Hence this can be concluded that the last rule in the delegation chain decides the path type
+//
 // pathMatchConditionsValid guarantees that if a prefix is present, it will start with a
 // / character, so we can simply concatenate.
-func mergePathMatchConditions(conds []contour_api_v1.MatchCondition) *PrefixMatchCondition {
-	prefix := ""
+func mergePathMatchConditions(conds []contour_api_v1.MatchCondition) MatchCondition {
```

Comment:
we'll definitely need some thorough unit tests to make sure this logic is well defined and has proper coverage

Reply:
> re the rules around merging path conditions in the tree of HTTPProxy includes:
> 
> ```
> // If the root condition is prefix, and included condition is prefix then after merging we get one prefix condition.
> // If the root condition is prefix, and included condition is exact then after merging we get one exact condition.
> // If the root condition is exact, and included condition is exact then after merging we get one exact condition.
> // If the root condition is exact, and included condition is prefix then after merging we get one prefix condition.
> ```
> 
> an alternate interpretation of the "exact" match in this context could be that such a match is terminal, i.e. it must come as a leaf of the tree of match conditions, since as it stands above if an exact match is at the root or middle of an include tree it effectively becomes a prefix match. if we want to keep these things distinct/clear that might be a way to go

Agreed @sunjayBhatia, If you see the next line ahead of these comments I have mentioned the interpretation. These comments are just to explain why we reached the conclusion that the leaf decides the path type.

```
// Hence this can be concluded that the last rule in the delegation chain decides the path type
```

Let me know if you meant something else.